TombEngine/Objects/Effects/Boss.cpp

```diff
@@ -0,0 +1,321 @@
+
+			if (item.Model.Color.x <= 0.0f &&
+				item.Model.Color.y <= 0.0f &&
+				item.Model.Color.z <= 0.0f)
+			{
+				RemoveActiveItem(itemNumber);
+				KillItem(itemNumber);
+			}
+		}
+
+		item.Model.Mutator[0].Scale += Vector3::One;
+		UpdateItemRoom(itemNumber);
+	}
+
+	void ShockwaveExplosionControl(int itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		// Entity life.
+		if (item.ItemFlags[0] > 0)
+			item.ItemFlags[0]--;
+
+		if (item.ItemFlags[0] <= 0)
+		{
+			auto colorEnd = Vector4(item.Model.Color.x - 0.1f, item.Model.Color.y - 0.1f, item.Model.Color.z - 0.1f, item.Model.Color.w);
+			item.Model.Color = Vector4::Lerp(item.Model.Color, colorEnd, 0.35f);
+
+			if (item.Model.Color.x <= 0.0f &&
+				item.Model.Color.y <= 0.0f &&
+				item.Model.Color.z <= 0.0f)
+			{
+				RemoveActiveItem(itemNumber);
+				KillItem(itemNumber);
+			}
+		}
+
+		item.Pose.Orientation.y += ANGLE(5.0f);
+		item.Model.Mutator[0].Scale += Vector3(0.5f);
+		UpdateItemRoom(itemNumber);
+	}
+
+	void SpawnExplosionSmoke(const Vector3& pos)
+	{
+		auto& spark = *GetFreeParticle();
+
+		spark.on = true;
+		spark.sR = 75;
+		spark.sG = 75;
+		spark.sB = 175;
+		spark.dR = 25;
+		spark.dG = 25;
+		spark.dB = 100;
+		spark.colFadeSpeed = 8;
+		spark.fadeToBlack = 64;
+		spark.life =
+		spark.sLife = Random::GenerateInt(96, 128);
+		spark.blendMode = BLEND_MODES::BLENDMODE_ADDITIVE;
+		spark.x = pos.x + Random::GenerateInt(-16, 16);
+		spark.y = pos.y + Random::GenerateInt(-16, 16);
+		spark.z = pos.z + Random::GenerateInt(-16, 16);
+		spark.xVel = ((GetRandomControl() & 0xFFF) - 2048) >> 2;
+		spark.yVel = GetRandomControl() - 128;
+		spark.zVel = ((GetRandomControl() & 0xFFF) - 2048) >> 2;
+		spark.friction = 6;
+		spark.flags = SP_SCALE | SP_DEF | SP_ROTATE | SP_EXPDEF;
+		spark.rotAng = GetRandomControl() & 0xFFF;
+
+		if (Random::TestProbability(1 / 2.0f))
+			spark.rotAdd = -16 - (GetRandomControl() & 0xF);
+		else
+			spark.rotAdd = (GetRandomControl() & 0xF) + 16;
+
+		spark.scalar = 3;
+		spark.gravity = -3 - (GetRandomControl() & 3);
+		spark.maxYvel = -8 - (GetRandomControl() & 3);
+		spark.dSize = (GetRandomControl() & 0x1F) + 256;
+		spark.sSize = spark.dSize / 2;
+		spark.size = spark.dSize / 2;
+	}
+
+	void ExplodeBoss(int itemNumber, ItemInfo& item, int deathCountToDie, const Vector4& color)
+	{
+		auto pos = Vector3::Zero;
+
+		// Disable shield.
+		item.SetFlagField(BOSSFlag_ShieldIsEnabled, 0);
+		item.HitPoints = NOT_TARGETABLE;
+
+		// Start doing the explosion (entity will do the count).
+		short counter = item.ItemFlags[BOSSFlag_ExplodeCount];
+		if (counter == 1)
+		{
+			SpawnShockwaveExplosion(item, color);
+			for (int i = 0; i < 3; i++)
+			{
+				pos = Vector3(
+					item.Pose.Position.x + Random::GenerateInt(-512, 512),
+					(item.Pose.Position.y - CLICK(3)) + Random::GenerateInt(-512, 512),
+					item.Pose.Position.z + Random::GenerateInt(-512, 512));
+				SpawnExplosionSmoke(pos);
+			}
+		}
+
+		if (counter == 10 ||
```

Comment:
Can we dehardcode this or make it check if `counter` can be divided by 10 without remainder (maybe using modulo)?